File: src/experimot_rave/KinectBodyHelper.cpp
#include "KinectBodyHelper.h"

KinectBodyHelper* KinectBodyHelper::_instance = NULL;

// Singleton Helper class
KinectBodyHelper* KinectBodyHelper::Instance(){
	if (!_instance){
		_instance = new KinectBodyHelper();
	}
	return _instance;
}
KinectBodyHelper::KinectBodyHelper()
{
	_init();
}
KinectBodyHelper::~KinectBodyHelper()
{
	if (_instance){
		delete _instance;
		_instance = NULL;
	}
}
const std::vector<std::tuple<KinectJoint_JointType, KinectJoint_JointType>>& KinectBodyHelper::GetBones() const{
	return _bonesList;
}
int KinectBodyHelper::GetBoneWidth() const{
	return _boneWidth;
}
void KinectBodyHelper::GetBodyColor(int index, RaveVector<float>& color){
	if (index >= 0 && index < _bodyColors.size())
		color = _bodyColors[index];
	else
		ColorHelper::HexToRgb(BLACK, color);
}
void KinectBodyHelper::_init(){

	// Initialize vars
	_boneWidth = 6;

	// Initialize Bones

	// Torso
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_Head, KinectJoint_JointType::KinectJoint_JointType_Neck));
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_Neck, KinectJoint_JointType::KinectJoint_JointType_SpineShoulder));
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_SpineShoulder, KinectJoint_JointType::KinectJoint_JointType_SpineMid));
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_SpineMid, KinectJoint_JointType::KinectJoint_JointType_SpineBase));
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_SpineShoulder, KinectJoint_JointType::KinectJoint_JointType_ShoulderRight));
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_SpineShoulder, KinectJoint_JointType::KinectJoint_JointType_ShoulderLeft));
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_SpineBase, KinectJoint_JointType::KinectJoint_JointType_HipRight));
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_SpineBase, KinectJoint_JointType::KinectJoint_JointType_HipLeft));

	// Right Arm
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_ShoulderRight, KinectJoint_JointType::KinectJoint_JointType_ElbowRight));
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_ElbowRight, KinectJoint_JointType::KinectJoint_JointType_WristRight));
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_WristRight, KinectJoint_JointType::KinectJoint_JointType_HandRight));
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_HandRight, KinectJoint_JointType::KinectJoint_JointType_HandTipRight));
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_WristRight, KinectJoint_JointType::KinectJoint_JointType_ThumbRight));

	// Left Arm
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_ShoulderLeft, KinectJoint_JointType::KinectJoint_JointType_ElbowLeft));
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_ElbowLeft, KinectJoint_JointType::KinectJoint_JointType_WristLeft));
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_WristLeft, KinectJoint_JointType::KinectJoint_JointType_HandLeft));
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_HandLeft, KinectJoint_JointType::KinectJoint_JointType_HandTipLeft));
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_WristLeft, KinectJoint_JointType::KinectJoint_JointType_ThumbLeft));

	// Right Leg
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_HipRight, KinectJoint_JointType::KinectJoint_JointType_KneeRight));
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_KneeRight, KinectJoint_JointType::KinectJoint_JointType_AnkleRight));
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_AnkleRight, KinectJoint_JointType::KinectJoint_JointType_FootRight));

	// Left Leg
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_HipLeft, KinectJoint_JointType::KinectJoint_JointType_KneeLeft));
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_KneeLeft, KinectJoint_JointType::KinectJoint_JointType_AnkleLeft));
	_bonesList.push_back(std::make_tuple(KinectJoint_JointType::KinectJoint_JointType_AnkleLeft, KinectJoint_JointType::KinectJoint_JointType_FootLeft));
	// Initialize Body Colors

	// populate body colors, one for each BodyIndex
	RaveVector<float> color;
	ColorHelper::HexToRgb(RED1, color);
	_bodyColors.push_back(color);
	ColorHelper::HexToRgb(ORANGE, color);
	_bodyColors.push_back(color);
	ColorHelper::HexToRgb(GREEN, color);
	_bodyColors.push_back(color);
	ColorHelper::HexToRgb(BLUE, color);
	_bodyColors.push_back(color);
	ColorHelper::HexToRgb(INDIGO, color);
	_bodyColors.push_back(color);
	ColorHelper::HexToRgb(VIOLET, color);
	_bodyColors.push_back(color);

	// Other colors
	ColorHelper::FromArgb(128, 255, 0, 0, handClosedBrush);
	ColorHelper::FromArgb(128, 0, 255, 0, handOpenBrush);
	ColorHelper::FromArgb(128, 0, 0, 255, handLassoBrush);
	ColorHelper::FromArgb(255, 68, 192, 68, trackedJointBrush);

	ColorHelper::HexToRgb(YELLOW1, inferredJointBrush); // = Brushes.Yellow;
	ColorHelper::HexToRgb(GRAY1, inferredBonePen);// = new Pen(Brushes.Gray, 1);
}

File: src/Experimot.Kinect.Perception/KinectBodyHelper.cs
using System;
using System.Collections.Generic;
using System.Windows.Media;
using Microsoft.Kinect;

namespace Experimot.Kinect.Perception
{
    public class KinectBodyHelper
    {
        private static KinectBodyHelper _instance;

        /// <summary>
        /// Radius of drawn hand circles
        /// </summary>
        public const double HandSize = 30;

        /// <summary>
        /// Thickness of drawn joint lines
        /// </summary>
        public const double JointThickness = 3;

        /// <summary>
        /// Thickness of clip edge rectangles
        /// </summary>
        public const double ClipBoundsThickness = 10;

        /// <summary>
        /// Constant for clamping Z values of camera space points from being negative
        /// </summary>
        public const float InferredZPositionClamp = 0.1f;

        /// <summary>
        /// Brush used for drawing hands that are currently tracked as closed
        /// </summary>
        public readonly Brush HandClosedBrush = new SolidColorBrush(Color.FromArgb(128, 255, 0, 0));

        /// <summary>
        /// Brush used for drawing hands that are currently tracked as opened
        /// </summary>
        public readonly Brush HandOpenBrush = new SolidColorBrush(Color.FromArgb(128, 0, 255, 0));

        /// <summary>
        /// Brush used for drawing hands that are currently tracked as in lasso (pointer) position
        /// </summary>
        public readonly Brush HandLassoBrush = new SolidColorBrush(Color.FromArgb(128, 0, 0, 255));

        /// <summary>
        /// Brush used for drawing joints that are currently tracked
        /// </summary>
        public readonly Brush TrackedJointBrush = new SolidColorBrush(Color.FromArgb(255, 68, 192, 68));

        /// <summary>
        /// Brush used for drawing joints that are currently inferred
        /// </summary>        
        public readonly Brush InferredJointBrush = Brushes.Yellow;

        /// <summary>
        /// Pen used for drawing bones that are currently inferred
        /// </summary>        
        public readonly Pen InferredBonePen = new Pen(Brushes.Gray, 1);

        /// <summary>
        /// definition of bones
        /// </summary>
        private readonly List<Tuple<JointType, JointType>> _bones;

        /// <summary>
        /// List of colors for each body tracked
        /// </summary>
        private readonly List<Pen> _bodyColors;


        private KinectBodyHelper()
        {
            _bones = new List<Tuple<JointType, JointType>>();
            // a bone defined as a line between two joints
            _bones = new List<Tuple<JointType, JointType>>
            {
                // Torso
                new Tuple<JointType, JointType>(JointType.Head, JointType.Neck),
                new Tuple<JointType, JointType>(JointType.Neck, JointType.SpineShoulder),
                new Tuple<JointType, JointType>(JointType.SpineShoulder, JointType.SpineMid),
                new Tuple<JointType, JointType>(JointType.SpineMid, JointType.SpineBase),
                new Tuple<JointType, JointType>(JointType.SpineShoulder, JointType.ShoulderRight),
                new Tuple<JointType, JointType>(JointType.SpineShoulder, JointType.ShoulderLeft),
                new Tuple<JointType, JointType>(JointType.SpineBase, JointType.HipRight),
                new Tuple<JointType, JointType>(JointType.SpineBase, JointType.HipLeft),
                // Right arm
                new Tuple<JointType, JointType>(JointType.ShoulderRight, JointType.ElbowRight),
                new Tuple<JointType, JointType>(JointType.ElbowRight, JointType.WristRight),
                new Tuple<JointType, JointType>(JointType.WristRight, JointType.HandRight),
                new Tuple<JointType, JointType>(JointType.HandRight, JointType.HandTipRight),
                new Tuple<JointType, JointType>(JointType.WristRight, JointType.ThumbRight),
                // Left arm
                new Tuple<JointType, JointType>(JointType.ShoulderLeft, JointType.ElbowLeft),
                new Tuple<JointType, JointType>(JointType.ElbowLeft, JointType.WristLeft),
                new Tuple<JointType, JointType>(JointType.WristLeft, JointType.HandLeft),
                new Tuple<JointType, JointType>(JointType.HandLeft, JointType.HandTipLeft),
                new Tuple<JointType, JointType>(JointType.WristLeft, JointType.ThumbLeft),
                // Right leg
                new Tuple<JointType, JointType>(JointType.HipRight, JointType.KneeRight),
                new Tuple<JointType, JointType>(JointType.KneeRight, JointType.AnkleRight),
                new Tuple<JointType, JointType>(JointType.AnkleRight, JointType.FootRight),
                // Left leg
                new Tuple<JointType, JointType>(JointType.HipLeft, JointType.KneeLeft),
                new Tuple<JointType, JointType>(JointType.KneeLeft, JointType.AnkleLeft),
                new Tuple<JointType, JointType>(JointType.AnkleLeft, JointType.FootLeft)
            };

            // populate body colors, one for each BodyIndex
            _bodyColors = new List<Pen>
            {
                new Pen(Brushes.Red, 6),
                new Pen(Brushes.Orange, 6),
                new Pen(Brushes.Green, 6),
                new Pen(Brushes.Blue, 6),
                new Pen(Brushes.Indigo, 6),
                new Pen(Brushes.Violet, 6)
            };

        }

        public static KinectBodyHelper Instance
        {
            get { return _instance ?? (_instance = new KinectBodyHelper()); }
        }

        /// <summary>
        /// definition of bones
        /// </summary>
        public List<Tuple<JointType, JointType>> Bones
        {
            get { return _bones; }
        }

        /// <summary>
        /// List of colors for each body tracked
        /// </summary>
        public List<Pen> BodyColors
        {
            get { return _bodyColors; }
        }
    }
}
